sim-lib/src/lib.rs

```diff
@@ -167,25 +159,33 @@ pub enum PaymentOutcome {
     IncorrectPaymentDetails,
     InsufficientBalance,
     Unknown,
+    NotDispatched,
+    TrackPaymentFailed,
 }
 
-#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
-struct DispatchedPayment {
+/// Describes a payment from a source node to a destination node.
+#[derive(Debug, Clone, Copy, Serialize)]
+struct Payment {
+    /// Pubkey of the source node dispatching the payment.
     source: PublicKey,
+    /// Pubkey of the destination node receiving the payment.
     destination: PublicKey,
-    #[serde(with = "serializers::serde_payment_hash")]
-    hash: PaymentHash,
+    /// Amount of the payment in msat.
     amount_msat: u64,
+    /// Hash of the payment if it has been successfully dispatched.
+    #[serde(with = "serializers::serde_option_payment_hash")]
+    hash: Option<PaymentHash>,
+    /// Time at which the payment was dispatched.
     #[serde(with = "serde_millis")]
     dispatch_time: SystemTime,
 }
 
-impl Display for DispatchedPayment {
+impl Display for Payment {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         write!(
             f,
             "Payment {} dispatched at {:?} sending {} msat from {} -> {}",
-            hex::encode(self.hash.0),
+            hex::encode(self.hash.map(|h| hex::encode(h.0)).unwrap_or(String::new())),
```

Comment:
This is still redundant. You're hex-encoding it twice.

```
self.hash.map(|h| hex::encode(h.0)).unwrap_or(String::new()),
```

Reply:
oh I didn't catch it on previous pass